# Étude à l'échelle de la consommation par adresse

Nous tâchons désormais de savoir s'il est possible pour un producteur d'électricité de prévoir la consommation d'électricité *d'une résidence* à partir d'un autre type de données plus local que des données départementales : des données à l'échelle des communes et des adresses. Nous continuons à considérer la météo comme une variable potentiellement pertinente pour comprendre la consommation, et nous cherchons à ajouter à cela la performance énergétique des bâtiments en passant par l'indice de DPE ([Diagnostic de performance énergétique](https://fr.wikipedia.org/wiki/Diagnostic_de_performance_%C3%A9nerg%C3%A9tique)). Ainsi, on cherche à savoir si les producteurs d'électricité, connaissant la notation énergétique d'un bâtiment toutes énergies comprises peuvent en prédire efficacement la consommation électrique de cette adresse au cours de l'année à venir.

Ce notebook s'organise en deux temps. 

- **Premier temps** : Récupération des données et formation d'un data frame regroupant toutes les données intéressantes.   
  Ce premier temps s'organise de la manière suivante :

      1. Récupération des données de consommation électriques individuelles sur le site d'Enedis
      2. Récupération des données météorologiques sur le site d'Infoclimat à l'échelle de communes, at agrégation de ces données pour constituer des moyennes annuelles
      3. Introduction d'une base de données de l'INSEE pour normaliser les noms de commune de la base de données météorologiques
      4. Constitution d'une liste de communes d'étude
      5. Récupération des données DPE
      6. Fusion des différentes données   

    
- **Second temps** : Régression utiles et conclusions

## Introduction - Mise en route

Le plan est le suivant : on cherche à régresser la consommation à l'échelle des adresses en fonction de la météo et de l'indice DPE de l'habitation. Nous cherchons donc une base de données d'adresses avec ces trois caractéristiques. Pour cela, nous allons checher à fusionner trois différentes bases de données :
- Des bases de données de consommations annuelles par adresse sur le site d'[Enedis](https://enedis.opendatasoft.com/explore/dataset/consommation-annuelle-residentielle-par-adresse/information/)
- Des bases de données météorologiques à l'échelle des communes sur le site d'[Infoclimat](https://www.infoclimat.fr/stations-meteo/analyses-mensuelles.php?mois)
- Des bases de données de notation DPE par adresse fournie par l'[ADEME (Agence de l'environnement et de la maîtrise de l'énergie)](https://data.ademe.fr/datasets/dpe-france)

Pour permettre la bonne fusion de ces tables, nous importons en plus une base de données de l'INSEE via le module `cartiflette`, pour permettre d'associer à chaque commune son nom officiel et son numéro INSEE de commune.

Les bâtiments sur lesquels portera notre étude seront donc les résidences localisées dans des communes proposant des données météorologiques, et ayant renseignés à la fois leur indice DPE et leur consommation électrique annuelle.

On commence par installer puis importer toutes les librairies et tous les modules qui nous seront utiles.

In [17]:
!pip install pandas
!pip install geopandas
!pip install geoviews
!pip install lxml
!pip install urllib
!pip install matplotlib
!pip install requests py7zr geopandas openpyxl tqdm s3fs PyYAML xlrd
!pip install git+https://github.com/inseefrlab/cartiflette@80b8a5a28371feb6df31d55bcc2617948a5f9b1a
!pip install mapclassify
!pip install folium

ERROR: Could not find a version that satisfies the requirement urllib (from versions: none)
ERROR: No matching distribution found for urllib
  Cloning https://github.com/inseefrlab/cartiflette (to revision 80b8a5a28371feb6df31d55bcc2617948a5f9b1a) to /tmp/pip-req-build-ocgnuvvs
  Running command git clone --filter=blob:none --quiet https://github.com/inseefrlab/cartiflette /tmp/pip-req-build-ocgnuvvs
  Running command git rev-parse -q --verify 'sha^80b8a5a28371feb6df31d55bcc2617948a5f9b1a'
  Running command git fetch -q https://github.com/inseefrlab/cartiflette 80b8a5a28371feb6df31d55bcc2617948a5f9b1a
  Running command git checkout -q 80b8a5a28371feb6df31d55bcc2617948a5f9b1a
  Resolved https://github.com/inseefrlab/cartiflette to commit 80b8a5a28371feb6df31d55bcc2617948a5f9b1a
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done


In [18]:
!pip install -q lxml

In [19]:
from declarations import *

In [20]:
import bs4
import lxml
import pandas as pd
import urllib
from tqdm.auto import tqdm

from urllib import request

## 1. Construction d'un tableau de données utile à notre étude 

### 1.1. Récupération des données de consommation électrique individuelle

Pour une étude sur la consommation électrique des bâtiments résidentiels, il paraît raisonnable de commencer par importer des donnés de consommation électrique. À cette fin, on utilise la fonction get_data_consumption, codée dans la page declarations.py , pour récupérer les données de consommation par adresse d'Enedis à partir d'URL, et ce pour quatre années différentes.

In [21]:
import declarations as dec

df_cons_2018=dec.get_data_consumption(dec.consumption_data_url_2018, "2018")
df_cons_2019=dec.get_data_consumption(dec.consumption_data_url_2019, "2019")
df_cons_2020=dec.get_data_consumption(dec.consumption_data_url_2020, "2020")
df_cons_2021=dec.get_data_consumption(dec.consumption_data_url_2021, "2021")

Chargement des données, cette étape peut prendre quelques minutes
Téléchargement réussi.
Chargement des données, cette étape peut prendre quelques minutes
Téléchargement réussi.


/home/onyxia/work/Projet_Python/declarations.py:321: DtypeWarning: Columns (16) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv(path_to_data, sep=";")


Chargement des données, cette étape peut prendre quelques minutes
Téléchargement réussi.


/home/onyxia/work/Projet_Python/declarations.py:321: DtypeWarning: Columns (16) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv(path_to_data, sep=";")


Chargement des données, cette étape peut prendre quelques minutes
Téléchargement réussi.


On affiche ensuite l'une des tables pour montrer ce à quoi ressemblent nos données. On prend l'exemple de l'année 2021.

In [22]:
df_cons_2021

Année  Code IRIS                 Nom IRIS  Numéro de voie  \
0        2021  800010101   Centre Ville Émonville            43.0   
1        2021  800010502     Delique-Saint-Gilles             2.0   
2        2021  800010401  Faubourg de la Bouvaque            30.0   
3        2021  800010502     Delique-Saint-Gilles             8.0   
4        2021  800010502     Delique-Saint-Gilles            20.0   
...       ...        ...                      ...             ...   
407048   2021  751072705        École Militaire 5            15.0   
407049   2021  751072705        École Militaire 5            14.0   
407050   2021  682240903        Coteaux Sud-Ouest             8.0   
407051   2021  682240601            Nordfeld Nord            84.0   
407052   2021  751166130               Auteuil 30            10.0   

       Indice de répétition Type de voie     Libellé de voie  \
0                       NaN          RUE  BOUCHER DE PERTHES   
1                       NaN          RUE            D ARTOIS   
2                       NaN          RUE    DE LA CITE LEDAY   
3                       NaN          RUE        DE NORMANDIE   
4                       NaN          RUE        DE NORMANDIE   
...                     ...          ...                 ...   
407048                  NaN       AVENUE         DE BRETEUIL   
407049                  NaN       AVENUE         DE BRETEUIL   
407050                  NaN          RUE         JULES VERNE   
407051                  NaN          RUE             LAURENT   
407052                  NaN       AVENUE       DE VERSAILLES   

        Code INSEE de la commune Nom de la commune Segment de client  \
0                          80001         Abbeville       RESIDENTIEL   
1                          80001         Abbeville       RESIDENTIEL   
2                          80001         Abbeville       RESIDENTIEL   
3                          80001         Abbeville       RESIDENTIEL   
4                          80001         Abbeville       RESIDENTIEL   
...                          ...               ...               ...   
407048                     75056             Paris       RESIDENTIEL   
407049                     75056             Paris       RESIDENTIEL   
407050                     68224          Mulhouse       RESIDENTIEL   
407051                     68224          Mulhouse       RESIDENTIEL   
407052                     75056             Paris       RESIDENTIEL   

        Nombre de logements  Consommation annuelle totale de l'adresse (MWh)  \
0                        11                                           59.439   
1                        31                                          115.704   
2                        12                                           36.454   
3                        10                                           25.769   
4                        10                                           22.995   
...                     ...                                              ...   
407048                   16                                           55.216   
407049                   17                                           61.842   
407050                   15                                           23.942   
407051                   12                                           32.852   
407052                   11                                           45.091   

        Consommation annuelle moyenne par logement de l'adresse (MWh)  \
0                                                   5.404               
1                                                   3.732               
2                                                   3.038               
3                                                   2.577               
4                                                   2.299               
...                                                   ...               
407048                                              3.451               


### 1.2. Scrapping des données météo communales

Une fois les données de consommation électrique récupérées, nous souhaitons nous concentrer sur des données météorologiques à l'échelle la plus locale qu'il nous est possible d'obtenir. Le site Info Climat propose justement des données météorologiques relevées dans plusieurs stations françaises, à une fréquence quotidienne. Par simplicté, nous supposerons que les rélevés météorologiques d'une station portent sur toute la commune dans laquelle est placée la station. 

Les données que nous récupérons sont des moyennes mensuelles des relevés quotidiens. Cependant, nous n'avons accès qu'à des données de consommations individuelles *annuelles*. Pour donner du sens à ces données météorologiques dans notre étude, nous effectuons donc des moyennes annuelles à partir de ces données mensuelles. 

En régressant la consommation par des données météo d'un pas de temps aussi large, on se demande si le "contexte" climatique d'une région ou d'une époque influe sur la consommation d'électricité des ménages. Ce lien pourrait avoir du sens car intuitivement un climat plus froid demande plus de chauffage, et un environnement plus sombre demande plus d'écairage. 

Pour ordonner les différentes fonctions qui se rapportent à l'extraction de données météorologiques à l'échelle des communes, on se propose de construire une classe regroupant les différentes fonctions qui nous intéressent.

In [131]:
class Meteo :
    """
    Class permettant de scrapper et d'effectuer des opérations sur les données d'Infoclimat concernant la température observée par différentes 
    stations météo.
    """
    def __init__(self):
        """
        Initialisation de la class contenant :
            self.année_début : int
                Année de début des données qu'on souhaite scrapper
            self.année_fin : int
                Année de fin des données qu'on souhaite scrapper
            self.data_all : dict
                Contient l'ensemble des données par année (initialement vides)
        """
        #Par défaut, on cadre notre extraction de données entre 2018 et 2021, années pour lesquelles nous avons des données de consommation
        self.année_début=2018 
        self.année_fin=2021
        self.data_all = {i : {j : {} for j in range(1,13)} for i in range(self.année_début,self.année_fin+1)}

    def scrap(self):
        """
        Fonction permettant de scrapper les données voulues, i.e. les données d'infoclimat concernant la température observée par différentes 
        stations météo entre 2011 et 2018, avec un pré-tri permettant d'éviter toute répétition de ville au sein d'un même mois.

            Paramètres:
            ------------
                self.data_all : dict
                    Dictionnaire contenant toutes les données par années
        
            Output:
            ----------
                self.data_pop : dict
                    Dictionnaire contenant toutes les données par années actualisées
            
        """
        #On boucle sur toutes les années souhaitées
        for y in self.data_all :
            #On boucle ensuite sur tous les mois de l'année
            for m in self.data_all[y] :
                url_temp = 'https://www.infoclimat.fr/stations-meteo/analyses-mensuelles.php?mois=' + str(m) + '&annee=' + str(y)
                raw_text = request.urlopen(url_temp).read()
                page = bs4.BeautifulSoup(raw_text,'lxml')
                tableau = page.find('table', {'id' : 'tableau-releves'})
                rows = tableau.find_all('tr')
                
                #On stocke les données d'un mois et d'une année dans un dictionnaire
                df = {
            'ville' : [rows[i].find('a').text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text],
            'tnn' : [rows[i].find('div').text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text],
            'tnm' : [rows[i].find_all('td')[2].text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text],
            'tmm' : [rows[i].find_all('td')[3].text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text],
            'txm' : [rows[i].find_all('td')[4].text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text],
            'txx' : [rows[i].find_all('td')[5].find('div').text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text],
            'rr' : [rows[i].find_all('td')[6].text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text],
            "ens" : [rows[i].find_all('td')[7].text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text],
            'rafale' : [rows[i].find_all('td')[8].text for i in range(1,len(rows)) if rows[i].find('a').text!=rows[i-1].find('a').text] 
                    }
                #La condition "if" dans toutes les listes permet de nous assurer que nous ne prenons qu'une seule donnée par ville
                #Par défaut, si cette ville est renseignée deux fois, nous ne prenons que la première information
                #Ceci nous permet d'éviter les répétitions de villes, qui bruitent la suite
               
                self.data_all[y][m] = df

    def export(self):
        """
        Fonction intégrée d'export au format DataFrame et au format .json
        pour permettant de faciliter les manipulations.
        Cette méthode ne sera pas utilisée en pratique dans notre projet

        Paramètres:
        ------------
            self.data_pop : dict
                Dictionnaire contenant toutes les données par années
        
        Output:
        ----------
            Aucun
        """
        
        self.df_all = pd.DataFrame(self.data_all)
        self.df_all.to_json('data_base.json')

    def liste_min(self, year):
        """
        Comme les stations ne sont pas les mêmes d'un mois à l'autre,
        cette fonction permet, pour une année donnée, de construire la liste des stations communes 
        à tous les mois pour une année donnée.

        Entrée:
        ------------
            year : int
                Année qui nous intéresse. Elle doit être entre self.année_début=2018 et self.année_fin

        Paramètres:
        ------------
            self.data_all : dict
                Dictionnaire contenant toutes les données par années
        
        Output:
        ----------
            intersection : list
                liste des stations communes à tous les mois pour une année donnée.
        """
        intersection=self.data_all[year][1]['ville']
        for m in range(2,13):
            intersection=[x for x in intersection if x in self.data_all[year][m]['ville']]
        return intersection

    def dico_minimal(self,year) :
        """
        Cette fonction construit le dictionnaire des données pour chaque mois d'une année "year" donnée, 
        mais seulement pour les villes proposant des données pour tous les mois de l'année.

        Entrée:
        ------------
            year : int
                Année qui nous intéresse. Elle doit être entre self.année_début=2018 et self.année_fin

        Paramètres:
        ------------
            self.data_all : dict
                Dictionnaire contenant toutes les données par années
            self.list_min : function
                Méthode de la classe Meteo qui renvoie la liste des villes communes à tous les mois 
                pour une année donnée
        
        Output:
        ----------
            df_min : dict
                dictionnaire des données pour chaque mois d'une année "year" donnée, 
                mais seulement pour les villes pour lesquelles on a des données pour tous les mois
        """
        intersection = self.liste_min(year)
        df_min={j : {} for j in range(1,13)}
        for m in range (1,13):
            df={'ville':[],'tnn':[],'tnm':[],'tmm':[],'txm':[],'txx':[],'rr':[],"ens":[],'rafale':[]}
            for i in range (len(self.data_all[year][m]['ville'])):
                if self.data_all[year][m]['ville'][i] in intersection and self.data_all[year][m]['ville'][i] != self.data_all[year][m]['ville'][i-1]: 
                    df['ville'].append(self.data_all[year][m]['ville'][i])
                    df['tnn'].append(self.data_all[year][m]['tnn'][i])
                    df['tnm'].append(self.data_all[year][m]['tnm'][i])
                    df['tmm'].append(self.data_all[year][m]['tmm'][i])
                    df['txm'].append(self.data_all[year][m]['txm'][i])
                    df['txx'].append(self.data_all[year][m]['txx'][i])
                    df['rr'].append(self.data_all[year][m]['rr'][i])
                    df['ens'].append(self.data_all[year][m]['ens'][i])
                    df['rafale'].append(self.data_all[year][m]['rafale'][i])
            df_min[m]=df       
        return df_min
        
        

    def tableau_annuel(self,year):
        """
        Fonction intégrée d'export en data frame des données météo sous forme de moyenne pour une année.
        Il faut que cette année "year" ait été importée, c'est-à-dire qu'elle soit 
        entre self.année_début et self.année_fin

        Paramètres:
        ------------
            self.dico_minimal(year) : dict
                Dictionnaire contenant toutes les données par mois pour les stations aux données fournies
                pour toute l'année
            self.dico_minimal(year) : list
                Liste des stations offrant des données pour toute l'année
        
        Output:
        ----------
            data : DataFrame
        """
        dico_min=self.dico_minimal(year)
        L=self.liste_min(year)
        df={j : {} for j in ['ville','tnn','tnm','tmm','txm','txx','rr','ens','rafale']}
        for x in range(len(L)):
            df['ville'][x]=L[x]
        for j in ['tnn','tnm','tmm','txm','txx','rr','ens','rafale']:
            dico={ville : [] for ville in L} 
            for x in range(len(L)):
                for m in range(1,13):
                    u=0
                    if dico_min[m][j][x]!='':
                        u+=1
                        dico[L[x]].append(float(dico_min[m][j][x]))
                    if u==0 :
                        moyenne='Rien'
                    else :
                        moyenne = sum(dico[L[x]])/u
                df[j][x]=moyenne
        data= pd.DataFrame(df)
        data.rename({'ville' : 'Villes', 
            'tnn' : 'Température minimale extrême du mois',
            'tnm' : 'Moyenne des températures minimales du mois', 
            'tmm' : 'Température moyenne du mois', 
            'txm' : 'Moyenne des températures maximales du mois', 
            'txx' : 'Température maximale extrême du mois', 
            'rr' : 'Cumul de précipitation du mois', 
            'ens' : "Heure d'ensoleillement du mois", 
            'rafale' : 'Rafale maximale du mois'},axis=1,inplace = True)
        return(data)
        

    def rename_pratique(self):
        """
        Fonction intégrée permettant de renommer facilement les colonnes par le nom utilisé par Infoclimat dans le but de faciliter le traitement.

        Paramètres :
        ------------
            self.df_all : DataFrame
                DataFrame contenant toutes les données météo extraites

        Output :
        --------
            self.df_all : DataFrame
                DataFrame contenant les mêmes données mais dont les colonnes ont été renommées
        """
        
        self.df_all = pd.DataFrame(self.data_all)
        self.df_all.rename({'Villes' : 'ville', 
           'Température minimale extrême du mois' : 'tnn',
           'Moyenne des températures minimales du mois' : 'tnm', 
           'Température moyenne du mois' : 'tmm', 
           'Moyenne des températures maximales du mois' : 'txm', 
           'Température maximale extrême du mois' : 'txx', 
           'Cumul de précipitation du mois' : 'rr', 
           "Heure d'ensoleillement du mois" : 'ens', 
           'Rafale maximale du mois' : 'rafale'},axis=1,inplace = True)

    def rename_esthétique(self):
        """
        Fonction intégrée permettant de renommer facilement les colonnes par leur véritable nom, après les avoir simplifié.

        Paramètres :
        ------------
            self.df_all : DataFrame
                DataFrame contenant toutes les données météo extraites

        Output :
        --------
            self.df_all : DataFrame
                DataFrame contenant les mêmes données mais dont les colonnes ont été renommées
        """
        
        self.df_all = pd.DataFrame(self.data_all)
        self.df_all.rename({'ville' : 'Villes', 
            'tnn' : 'Température minimale extrême du mois',
            'tnm' : 'Moyenne des températures minimales du mois', 
            'tmm' : 'Température moyenne du mois', 
            'txm' : 'Moyenne des températures maximales du mois', 
            'txx' : 'Température maximale extrême du mois', 
            'rr' : 'Cumul de précipitation du mois', 
            'ens' : "Heure d'ensoleillement du mois", 
            'rafale' : 'Rafale maximale du mois'},axis=1,inplace = True)

    

In [23]:
import declarations
#Importation de la class Meteo
lameteo=declarations.Meteo()
#On récupère les données
lameteo.scrap()
#On exporte au format DataFrame
lameteo.export()

**Remarque** : On présente ici une rapide parenthèse pour montrer à quel point il est important d'éviter les répétitions de villes et de s'assurer que la station renseigne bien les données météo pour tous les mois de l'année. En effet, dans la cellule ci-dessous, on voit que lorsque qu'on demande le nombre de villes (donc le nombre de stations) avec des informations pour chaque mois, et bien il varie. C'est le cas pour 2019, mais aussi pour toutes les autres années d'études.

In [24]:
[len(lameteo.df_all[2019][i]['ville']) for i in range(1,13)] #Exemple de l'année 2019

[888, 884, 959, 961, 969, 969, 970, 973, 978, 981, 988, 991]

In [25]:
[len(lameteo.df_all[2020][i]['ville']) for i in range(1,13)] #Exemple de l'année 2020

[980, 1432, 1329, 1292, 1320, 1327, 1323, 1329, 1341, 1338, 1346, 1359]

Maintenant que la classe est créée, nous pouvons exporter les données pour voir ce à quoi elles ressemblent. Comme pour les données de consommation, on prend l'exemple de l'année 2021. 

In [27]:
meteo_2021=lameteo.tableau_annuel(2021)
meteo_2021

Villes  \
0                                    Abbeville (80)   
1                          Abbeville - centre  (80)   
2                                      ACHERES (78)   
3                                        ADAST (65)   
4                                         Agde (34)   
...                                             ...   
1286  [MAE] Lycée Emile Zola - AIX EN PROVENCE (13)   
1287    [MAE] Lycée Marguerite Filhol - FUMEL  (47)   
1288      [MAE] Lycée Parc de Vilgénis - MASSY (91)   
1289   [MAE] Lycée Saint Exupéry - LA ROCHELLE (17)   
1290        [MAE] Lycée Victor Louis - TALENCE (33)   

     Température minimale extrême du mois  \
0                                    23.6   
1                                    14.9   
2                                    -4.3   
3                                    14.7   
4                                    79.6   
...                                   ...   
1286                                 29.9   
1287                                 27.0   
1288                                 22.9   
1289                                 55.2   
1290                                 53.0   

     Moyenne des températures minimales du mois Température moyenne du mois  \
0                                          88.7                       133.5   
1                                          87.6                       136.1   
2                                          79.3                       140.0   
3                                          82.8                       151.4   
4                                         143.3                       192.1   
...                                         ...                         ...   
1286                                       70.6                       131.8   
1287                                      102.3                       160.8   
1288                                       93.5                       147.3   
1289                                       Rien                        Rien   
1290                                      124.9                       179.4   

     Moyenne des températures maximales du mois  \
0                                         177.5   
1                                         184.5   
2                                         200.3   
3                                         219.3   
4                                         240.8   
...                                         ...   
1286                                      192.9   
1287                                      219.4   
1288                                      202.1   
1289                                       Rien   
1290                                      236.2   

     Température maximale extrême du mois Cumul de précipitation du mois  \
0                                   264.8                           17.4   
1                                   270.0                          695.8   
2                                   296.2                          696.0   
3                                   325.0                          841.5   
4                                   314.3                          325.2   
...                                   ...                            ...   
1286                                321.6                          576.4   
1287                                311.5                          891.2   
1288                                305.5                           Rien   
1289                                163.2                          447.2   
1290                                330.6                          415.8   

     Heure d'ensoleillement du mois Rafale maximale du mois  
0                              Rien                    Rien  
1                              Rien                   460.2  
2                              Rien                    Rien  
3                              Rien                    Rien  
4                              Rien                   633.

### 1.3. Utilisation d'une base intermédiaire pour normaliser les noms de communes

Dans le tableau des données météo précédent, on remarque que certains noms de stations correspondent bien à des noms de ville, mais ça n'est pas le cas de tous. Pour simplifier le croisement avec les autres tableaux de données, nous allons donc importer un tableau de l'INSEE, permettant d'associer quand c'est possible le nom d'une station avec le code INSEE de la commune ainsi que ses données géographiques.

In [28]:
communes = s3.download_vectorfile_url_all(
    values = "metropole",
    crs = 4326,
    borders = "COMMUNE",
    vectorfile_format="topojson",
    filter_by="FRANCE_ENTIERE",
    source="EXPRESS-COG-CARTO-TERRITOIRE",
    year=2022)

communes["area"] = dep.to_crs(2154).area

https://minio.lab.sspcloud.fr/projet-cartiflette/diffusion/shapefiles-test1/year=2022/administrative_level=COMMUNE/crs=4326/FRANCE_ENTIERE=metropole/vectorfile_format=topojson/provider=IGN/source=EXPRESS-COG-CARTO-TERRITOIRE/raw.topojson


Downloading: : 70.4MiB [00:01, 56.1MiB/s]


In [29]:
communes

id  ID                       NOM  \
0      COMMUNE_0000000009754033 NaN                Connangles   
1      COMMUNE_0000000009760784 NaN                   Vidouze   
2      COMMUNE_0000000009742077 NaN                 Fouesnant   
3      COMMUNE_0000000009735245 NaN              Plougrescant   
4      COMMUNE_0000000009752504 NaN                 Montcarra   
...                         ...  ..                       ...   
34821  COMMUNE_0000000009742349 NaN                      Brie   
34822  COMMUNE_0000000009742038 NaN                   Orgères   
34823  COMMUNE_0000000009734653 NaN         Castillon-en-Auge   
34824  COMMUNE_0000000009734657 NaN      Saint-Pierre-en-Auge   
34825  COMMUNE_0000000009739242 NaN  Saint-Martin-de-Bossenay   

                          NOM_M INSEE_COM          STATUT  POPULATION  \
0                    CONNANGLES     43076  Commune simple         137   
1                       VIDOUZE     65462  Commune simple         243   
2                     FOUESNANT     29058  Commune simple        9864   
3                  PLOUGRESCANT     22218  Commune simple        1166   
4                     MONTCARRA     38250  Commune simple         569   
...                         ...       ...             ...         ...   
34821                      BRIE     35041  Commune simple        1006   
34822                   ORGERES     35208  Commune simple        5152   
34823         CASTILLON-EN-AUGE     14141  Commune simple         163   
34824      SAINT-PIERRE-EN-AUGE     14654  Commune simple        7329   
34825  SAINT-MARTIN-DE-BOSSENAY     10351  Commune simple         366   

      INSEE_CAN INSEE_ARR INSEE_DEP INSEE_REG SIREN_EPCI  \
0            11         1        43        84  200073419   
1            13         3        65        76  200072106   
2            11         4        29        53  242900660   
3            27         3        22        53  200065928   
4            24         2        38        84  200068542   
...         ...       ...       ...       ...        ...   
34821        12         1        35        53  243500634   
34822        12         3        35        53  243500139   
34823        19         3        14        28  200069532   
34824        18         3        14        28  200069532   
34825        11         2        10        44  241000488   

                                 source territoire  \
0      IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
1      IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
2      IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
3      IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
4      IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
...                                 ...        ...   
34821  IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
34822  IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
34823  IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
34824  IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   
34825  IGN:EXPRESS-COG-CARTO-TERRITOIRE  metropole   

                                                geometry          area  
0      POLYGON ((3.61421 45.27206, 3.61377 45.27255, ...  5.774291e+09  
1      POLYGON ((-0.02485 43.43002, -0.02500 43.43012...  7.418552e+09  
2      MULTIPOLYGON (((-3.97912 47.70392, -3.97897 47...  4.033539e+09  
3      MULTIPOLYGON (((-3.20273 48.84902, -3.20259 48...  4.720673e+09  
4      POLYGON ((5.43220 45.61477, 5.43220 45.61438, ...  7.365673e+09  
...                                                  ...           ...  
34821  POLYGON ((-1.58072 47.94312, -1.58057 47.94420...           NaN  
34822  POLYGON ((-1.64908 47.96488, -1.65011 47.96595...           NaN  
34823  POLYGON ((0.09085 49.04872, 0.09129 49.04872, ...           NaN  
34824  POLYGON ((0.10231 49.02482, 0.10246 49.02443, ...           NaN  
34825  POLYGON ((3.65935 48.43158, 3.65552 48.43227, ...           NaN  

[34826 rows x 16 columns]

On donne la liste des colonnes ainsi que le nombres d'éléments distincts que chacune contient pour se faire une idée plus précise de la base.

In [30]:
communes.nunique()

id            34826
ID                0
NOM           32596
NOM_M         32537
INSEE_COM     34826
STATUT            5
POPULATION     5786
INSEE_CAN        56
INSEE_ARR         9
INSEE_DEP        96
INSEE_REG        13
SIREN_EPCI     1244
source            1
territoire        1
geometry      34826
area             96
dtype: int64

### 1.4. Création d'une liste de villes d'étude

On garde pour la suite le nom des stations utiles. Ces noms nous permettront de filtrer les communes dont nous devons étudier les DPE.

On voit que les noms de stations ne sont pas très propres : ils incluent les numéros de département, voire incluent des quartiers ou regroupent plusieurs villes. On cherche donc à nettoyer ces noms, pour qu'ils puissent être comparés aux noms de communes donnés par l'API utile pour les DPE. 

On ne touche pour cela qu'à la liste "Villes" du data frame. Nous remplacerons ensuite la base initiale par la base nettoyée, et supprimons les lignes qui ne nous intéressent pas dans le data frame après nettoyage. 

L'API des DPE qui nous intéresse n'identifie une commune que par son code INSEE. Il faut donc que, dans la base de données de chaque station, je réussisse à fusionner la base météo avec une base de données qui comprend les codes communes INSEE à partir de la commune de la station.

Par simplicité, nous ne garderons que les stations dont le nom est un nom de commune officiel ; nous ne chercherons pas pour chaque station la commune à laquelle elle se rapporte.

In [41]:
#On ne garde que les stations dans les villes et pas celles de stations de "Météo à l'école" (MAE)
def sans_mae(year):
    '''Cette fonction supprime les listes du data frame correspondant aux stations MAE
    Argument : year (type : int, année voulue) 
    Sortie : L[~L['Villes'].str[0].isin(['['])] (type : DataFrame)'''
    L=lameteo.tableau_annuel(year)
    return L[~L['Villes'].str[0].isin(['['])]
#meteo_bis_2021=meteo_2021[~meteo_2021['Villes'].str[0].isin(['['])]
meteo_bis_2021=sans_mae(2021)
meteo_bis_2021

Villes  \
0                              Abbeville (80)   
1                    Abbeville - centre  (80)   
2                                ACHERES (78)   
3                                  ADAST (65)   
4                                   Agde (34)   
...                                       ...   
1255                    Xonrupt-Longemer (88)   
1256          Xonrupt-Longemer - Le Poli (88)   
1257  Xonrupt-Longemer - Refuge du Sotré (88)   
1258                      Yvoy-le-Marron (41)   
1259                              Yzeron (69)   

     Température minimale extrême du mois  \
0                                    23.6   
1                                    14.9   
2                                    -4.3   
3                                    14.7   
4                                    79.6   
...                                   ...   
1255                                -35.8   
1256                                -27.1   
1257                                -38.2   
1258                                 -3.5   
1259                                 10.0   

     Moyenne des températures minimales du mois Température moyenne du mois  \
0                                          88.7                       133.5   
1                                          87.6                       136.1   
2                                          79.3                       140.0   
3                                          82.8                       151.4   
4                                         143.3                       192.1   
...                                         ...                         ...   
1255                                       42.5                        95.5   
1256                                       47.3                        95.5   
1257                                       37.8                        68.7   
1258                                       76.9                       137.5   
1259                                       72.9                       120.4   

     Moyenne des températures maximales du mois  \
0                                         177.5   
1                                         184.5   
2                                         200.3   
3                                         219.3   
4                                         240.8   
...                                         ...   
1255                                      148.6   
1256                                      143.5   
1257                                       99.2   
1258                                      197.7   
1259                                      167.4   

     Température maximale extrême du mois Cumul de précipitation du mois  \
0                                   264.8                           17.4   
1                                   270.0                          695.8   
2                                   296.2                          696.0   
3                                   325.0                          841.5   
4                                   314.3                          325.2   
...                                   ...                            ...   
1255                                255.1                         1428.2   
1256                                249.1                         1906.5   
1257                                207.2                           Rien   
1258                                298.0                           85.6   
1259                                264.2                         1002.6   

     Heure d'ensoleillement du mois Rafale maximale du mois  
0                              Rien                    Rien  
1                              Rien                   460.2  
2                              Rien                    Rien  
3                              Rien                    Rien  
4                              Rien                   633.9  
...                             ...                     ...  


On voit qu'en plus par ville il peut il y avoir plusieurs stations (correspondant à plusieurs quartiers). On va donc essayer de supprimer les parties qui ne dénotent pas de la ville dans les noms de stations.

In [32]:
def nettoyage(liste):
    '''Cette fonction est censée séparer les numéros de département des villes dans une liste présentée comme la liste liste_communes
    Argument : liste (type : list ; liste de string) 
    Sortie : List (type : list ; liste de listes)'''
    List=[]
    for ville in liste:
        i=0
        mot=''
        while i < len(ville) and ville[i+1] != '(' :
            mot+=ville[i]
            i+=1
        mot=mot.upper()
        List.append(mot)
    
    return List

In [33]:
meteo_bis_2021['Villes']=pd.Series(nettoyage(meteo_bis_2021['Villes'].to_list()))

/tmp/ipykernel_300/349770013.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  meteo_bis_2021['Villes']=pd.Series(nettoyage(meteo_bis_2021['Villes'].to_list()))


In [34]:
meteo_bis_2021['Villes']

0                                ABBEVILLE
1                      ABBEVILLE - CENTRE 
2                                  ACHERES
3                                    ADAST
4                                     AGDE
                       ...                
1255                      XONRUPT-LONGEMER
1256            XONRUPT-LONGEMER - LE POLI
1257    XONRUPT-LONGEMER - REFUGE DU SOTRÉ
1258                        YVOY-LE-MARRON
1259                                YZERON
Name: Villes, Length: 1260, dtype: object

In [35]:
meteo_bis_2021

Villes Température minimale extrême du mois  \
0                              ABBEVILLE                                 23.6   
1                    ABBEVILLE - CENTRE                                  14.9   
2                                ACHERES                                 -4.3   
3                                  ADAST                                 14.7   
4                                   AGDE                                 79.6   
...                                  ...                                  ...   
1255                    XONRUPT-LONGEMER                                -35.8   
1256          XONRUPT-LONGEMER - LE POLI                                -27.1   
1257  XONRUPT-LONGEMER - REFUGE DU SOTRÉ                                -38.2   
1258                      YVOY-LE-MARRON                                 -3.5   
1259                              YZERON                                 10.0   

     Moyenne des températures minimales du mois Température moyenne du mois  \
0                                          88.7                       133.5   
1                                          87.6                       136.1   
2                                          79.3                       140.0   
3                                          82.8                       151.4   
4                                         143.3                       192.1   
...                                         ...                         ...   
1255                                       42.5                        95.5   
1256                                       47.3                        95.5   
1257                                       37.8                        68.7   
1258                                       76.9                       137.5   
1259                                       72.9                       120.4   

     Moyenne des températures maximales du mois  \
0                                         177.5   
1                                         184.5   
2                                         200.3   
3                                         219.3   
4                                         240.8   
...                                         ...   
1255                                      148.6   
1256                                      143.5   
1257                                       99.2   
1258                                      197.7   
1259                                      167.4   

     Température maximale extrême du mois Cumul de précipitation du mois  \
0                                   264.8                           17.4   
1                                   270.0                          695.8   
2                                   296.2                          696.0   
3                                   325.0                          841.5   
4                                   314.3                          325.2   
...                                   ...                            ...   
1255                                255.1                         1428.2   
1256                                249.1                         1906.5   
1257                                207.2                           Rien   
1258                                298.0                           85.6   
1259                                264.2                         1002.6   

     Heure d'ensoleillement du mois Rafale maximale du mois  
0                              Rien                    Rien  
1                              Rien                   460.2  
2                              Rien                    Rien  
3                              Rien                    Rien  
4                              Rien                   633.9  
...                             ...                     ...  
1255                           Rien                   382.8  
1256                         1212.1                   346.0  
1257                        

On cherche à savoir combien de villes on est amenés à garder si l'on ne veut pas trop se casser la tête et ne garder que les noms "jolis" de stations, pour lesquels on peut facilement trouver le nom de la ville d'origine.

In [36]:
filtre=meteo_bis_2021.merge(communes, left_on='Villes', right_on='NOM_M')[['NOM_M','STATUT','POPULATION','INSEE_COM','geometry']]
filtre

NOM_M           STATUT  POPULATION INSEE_COM  \
0           ABBEVILLE  Sous-préfecture       22980     80001   
1             ACHERES   Commune simple       21529     78005   
2             ACHERES   Commune simple         380     18001   
3               ADAST   Commune simple         297     65001   
4                AGDE   Commune simple       29600     34003   
..                ...              ...         ...       ...   
751      WINTERSBOURG   Commune simple         269     57747   
752          XERTIGNY   Commune simple        2586     88530   
753  XONRUPT-LONGEMER   Commune simple        1515     88531   
754    YVOY-LE-MARRON   Commune simple         760     41297   
755            YZERON   Commune simple         992     69269   

                                              geometry  
0    POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  
1    POLYGON ((2.04953 48.95702, 2.04938 48.95770, ...  
2    POLYGON ((2.44015 47.27926, 2.44015 47.27965, ...  
3    POLYGON ((-0.06867 42.97044, -0.07087 42.97014...  
4    MULTIPOLYGON (((3.50101 43.26349, 3.50130 43.2...  
..                                                 ...  
751  POLYGON ((7.19741 48.77898, 7.19668 48.77917, ...  
752  POLYGON ((6.45837 48.04351, 6.45807 48.04126, ...  
753  POLYGON ((7.00967 48.04292, 7.00791 48.04263, ...  
754  POLYGON ((1.84782 47.58568, 1.84724 47.58529, ...  
755  POLYGON ((4.56364 45.70208, 4.56364 45.70257, ...  

[756 rows x 5 columns]

Après ce filtre, il nous reste donc 755 communes d'études, ce qui reste satisfaisant : on est rassuré que notre tri très simple n'ait pas enlevé trop de villes d'étude.

Maintenant, nous fusionnons cette base triée de données météo avec la base des données de consommation.

In [37]:
L=df_cons_2021['Nom de la commune'].to_list()
df_cons_2021['Nom de la commune']=pd.Series([x.upper() for x in L])
filtre_conso_2021 = filtre.merge(df_cons_2021, left_on="NOM_M", right_on="Nom de la commune")[['NOM_M','STATUT','INSEE_COM','geometry','Code IRIS', 'Nom IRIS', 'Nombre de logements', "Consommation annuelle totale de l'adresse (MWh)", "Consommation annuelle moyenne par logement de l'adresse (MWh)", "Consommation annuelle moyenne de la commune (MWh)", "Adresse"]]   
filtre_conso_2021

NOM_M           STATUT INSEE_COM  \
0             ABBEVILLE  Sous-préfecture     80001   
1             ABBEVILLE  Sous-préfecture     80001   
2             ABBEVILLE  Sous-préfecture     80001   
3             ABBEVILLE  Sous-préfecture     80001   
4             ABBEVILLE  Sous-préfecture     80001   
...                 ...              ...       ...   
35160  XONRUPT-LONGEMER   Commune simple     88531   
35161  XONRUPT-LONGEMER   Commune simple     88531   
35162  XONRUPT-LONGEMER   Commune simple     88531   
35163  XONRUPT-LONGEMER   Commune simple     88531   
35164            YZERON   Commune simple     69269   

                                                geometry  Code IRIS  \
0      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010101   
1      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010502   
2      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010401   
3      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010502   
4      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010502   
...                                                  ...        ...   
35160  POLYGON ((7.00967 48.04292, 7.00791 48.04263, ...  885310000   
35161  POLYGON ((7.00967 48.04292, 7.00791 48.04263, ...  885310000   
35162  POLYGON ((7.00967 48.04292, 7.00791 48.04263, ...  885310000   
35163  POLYGON ((7.00967 48.04292, 7.00791 48.04263, ...  885310000   
35164  POLYGON ((4.56364 45.70208, 4.56364 45.70257, ...  692690000   

                                    Nom IRIS  Nombre de logements  \
0                     Centre Ville Émonville                   11   
1                       Delique-Saint-Gilles                   31   
2                    Faubourg de la Bouvaque                   12   
3                       Delique-Saint-Gilles                   10   
4                       Delique-Saint-Gilles                   10   
...                                      ...                  ...   
35160  Xonrupt-Longemer (commune non irisée)                   52   
35161  Xonrupt-Longemer (commune non irisée)                   10   
35162  Xonrupt-Longemer (commune non irisée)                   14   
35163  Xonrupt-Longemer (commune non irisée)                   12   
35164            Yzeron (commune non irisée)                   13   

       Consommation annuelle totale de l'adresse (MWh)  \
0                                               59.439   
1                                              115.704   
2                                               36.454   
3                                               25.769   
4                                               22.995   
...                                                ...   
35160                                          110.382   
35161                                           20.775   
35162                                           85.812   
35163                                           37.800   
35164                                           41.780   

       Consommation annuelle moyenne par logement de l'adresse (MWh)  \
0                                                  5.404               
1                                                  3.732               
2                                                  3.038               
3                                                  2.577               
4                                                  2.299               
...                                                  ...               
35160                                              2.123               
35161                                              2.078               
35162                                              6.129               
35163                                              3.150               
35164                                              3.214               

       Consommation annuelle moyenne de la commune (MWh)  \
0                                                  3.546   
1    

In [55]:
len(filtre_conso_2021['NOM_M'].unique())

249

In [69]:
liste_communes=filtre_conso_2021['INSEE_COM'].unique()
len(liste_communes)

322

Attention ! On remarque ici que notre base de données comporte beaucoup de villes homonymes. Il faudrait donc bien faire attention au code INSEE / au code postal d'une ville pour s'assurer qu'on ne la confond pas avec une autre ville du même nom.

Notre étude portera donc au plus sur des immeubles issus de 249 communes différentes !

On essaiera par la suite de trouver un moyen de relier l'adresse de la base de données des DPE à l'adresse de la base qu'on vient dd définir. Pour cela, on cherchera à normaliser l'adresse de notre base.

In [78]:
filtre_conso_2021['Adresse normalisée']=filtre_conso_2021['Adresse']+ ' '+ filtre_conso_2021['INSEE_COM']+ ' '+ filtre_conso_2021['NOM_M']
filtre_conso_2021

NOM_M           STATUT INSEE_COM  \
0       ABBEVILLE  Sous-préfecture     80001   
1       ABBEVILLE  Sous-préfecture     80001   
2       ABBEVILLE  Sous-préfecture     80001   
3       ABBEVILLE  Sous-préfecture     80001   
4       ABBEVILLE  Sous-préfecture     80001   
...           ...              ...       ...   
25167  YSSINGEAUX  Sous-préfecture     43268   
25168  YSSINGEAUX  Sous-préfecture     43268   
25169  YSSINGEAUX  Sous-préfecture     43268   
25170  YSSINGEAUX  Sous-préfecture     43268   
25171      YZERON   Commune simple     69269   

                                                geometry  Code IRIS  \
0      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010501   
1      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010502   
2      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010503   
3      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010502   
4      POLYGON ((1.83974 50.07802, 1.83856 50.07744, ...  800010502   
...                                                  ...        ...   
25167  POLYGON ((4.09598 45.09519, 4.09598 45.09529, ...  432680102   
25168  POLYGON ((4.09598 45.09519, 4.09598 45.09529, ...  432680102   
25169  POLYGON ((4.09598 45.09519, 4.09598 45.09529, ...  432680102   
25170  POLYGON ((4.09598 45.09519, 4.09598 45.09529, ...  432680102   
25171  POLYGON ((4.56364 45.70208, 4.56364 45.70257, ...  692690000   

                          Nom IRIS  Nombre de logements  \
0                 La Porte au Bois                   28   
1             Delique-Saint Gilles                   36   
2               Ponthieu-Bagatelle                   10   
3             Delique-Saint Gilles                   10   
4             Delique-Saint Gilles                   10   
...                            ...                  ...   
25167                          Sud                   10   
25168                          Sud                   12   
25169                          Sud                   12   
25170                          Sud                   18   
25171  Yzeron (commune non irisée)                   13   

       Consommation annuelle totale de l'adresse (MWh)  \
0                                              134.006   
1                                               91.020   
2                                               18.327   
3                                               37.592   
4                                               33.586   
...                                                ...   
25167                                           12.318   
25168                                           10.403   
25169                                           11.531   
25170                                           17.056   
25171                                           33.372   

       Consommation annuelle moyenne par logement de l'adresse (MWh)  \
0                                                  4.786               
1                                                  2.528               
2                                                  1.833               
3                                                  3.759               
4                                                  3.359               
...                                                  ...               
25167                                              1.232               
25168                                              0.867               
25169                                              0.961               
25170                                              0.948               
25171                                              2.567               

       Consommation annuelle moyenne de la commune (MWh)  \
0                                                  3.522   
1                                                  3.522   
2                                                  3.522   
3                                                  3.522   
4          

### 1.5. Récupération des données de Diagnostics de Performance Energétique (DPE)

On commence par programmer une fonction qui permet de récupérer des données à partir d'une APL. Voici l'exemple de Lino lorsqu'on essaye la demande avec une seule ville.

In [7]:
code_commune = "01450"
size = 100
api_root = "https://koumoul.com/data-fair/api/v1/datasets/dpe-france/lines"
url_api = f"{api_root}?page=1&after=10&format=json&q_mode=simple&qs=code_insee_commune_actualise" + "%3A%22" + f"{code_commune}" + "%22" + f"&size={size}&select=" + "%2A&sampling=neighbors"

In [70]:
import requests
import pandas as pd
import geopandas as gpd

def get_dpe_from_url(api_url):
    '''Argument:
        api_url (string) : url de l'API voulue
    Sortie:
        pandas.DataFrame : dataframe
    '''
    req=requests.get(api_url)
    wb=req.json()
    df = pd.json_normalize(wb["results"])
    dpe = gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(df.longitude, df.latitude), crs = 4326)
    dpe = dpe.dropna(subset = ['longitude', 'latitude'])
    return dpe

dpe = get_dpe_from_url(url_api)
dpe

classe_consommation_energie tr001_modele_dpe_type_libelle  \
0                            D                         Vente   
2                            D                          Neuf   
3                            D                          Neuf   
6                            E                         Vente   
7                            E                         Vente   
..                         ...                           ...   
95                           C                      Location   
96                           B                      Location   
97                           B                      Location   
98                           C                      Location   
99                           E                      Location   

    annee_construction           _geopoint   latitude  surface_thermique_lot  \
0                 1947  45.925922,5.229964  45.925922                 117.16   
2                 2006  45.923421,5.223777  45.923421                  90.53   
3                 2006  45.923068,5.223271  45.923068                 105.27   
6                 1983   45.91956,5.224214  45.919560                 130.00   
7                 2006  45.916251,5.221846  45.916251                  90.34   
..                 ...                 ...        ...                    ...   
95                2019  45.920228,5.222557  45.920228                  54.82   
96                2019  45.920228,5.222557  45.920228                  65.89   
97                2019  45.920292,5.223036  45.920292                  67.77   
98                2019  45.920292,5.223036  45.920292                  58.23   
99                2019  45.920228,5.222557  45.920228                  35.40   

        _i tr002_type_batiment_description  \
0      487             Maison Individuelle   
2      689             Maison Individuelle   
3      690             Maison Individuelle   
6     1127             Maison Individuelle   
7     1374             Maison Individuelle   
..     ...                             ...   
95  102717                        Logement   
96  102718                        Logement   
97  102719                        Logement   
98  102720                        Logement   
99  102721                        Logement   

                                          geo_adresse   _rand  ...  \
0      Rue de la Brugnière 01800 Villieu-Loyes-Mollon   23215  ...   
2     Chemin du Pont-vieux 01800 Villieu-Loyes-Mollon  401672  ...   
3   Lotissement le Pont Vieux 01800 Villieu-Loyes-...  428929  ...   
6            Rue du Pollet 01800 Villieu-Loyes-Mollon  848537  ...   
7   6 Rue des Terres du Pollet 01800 Villieu-Loyes...  598270  ...   
..                                                ...     ...  ...   
95        25 Rue du Bottet 01800 Villieu-Loyes-Mollon  235935  ...   
96        25 Rue du Bottet 01800 Villieu-Loyes-Mollon  140973  ...   
97  257 Avenue Charles de Gaulle 01800 Villieu-Loy...  912454  ...   
98  257 Avenue Charles de Gaulle 01800 Villieu-Loy...  755526  ...   
99        25 Rue du Bottet 01800 Villieu-Loyes-Mollon  174306  ...   

   classe_estimation_ges  nom_methode_dpe  tv016_departement_code  \
0                      E  Méthode Facture                      01   
2                      C    FACTURE - DPE                      01   
3                      C    FACTURE - DPE                      01   
6                      C      Méthode 3CL                      01   
7                      C        3CL - DPE                      01   
..                   ...              ...                     ...   
95                     D        3CL - DPE                      01   
96                     C        3CL - DPE                      01   
97                     C        3CL - DPE                      01   
98                     D        3CL - DPE                      01   
99                     C        3CL - DPE                      01   

   consommation_energie date_etablissement_dpe lon

In [69]:
list(dpe)

['classe_consommation_energie',
 'tr001_modele_dpe_type_libelle',
 'annee_construction',
 '_geopoint',
 'latitude',
 'surface_thermique_lot',
 '_i',
 'tr002_type_batiment_description',
 'geo_adresse',
 '_rand',
 'code_insee_commune_actualise',
 'estimation_ges',
 'geo_score',
 'classe_estimation_ges',
 'nom_methode_dpe',
 'tv016_departement_code',
 'consommation_energie',
 'date_etablissement_dpe',
 'longitude',
 '_score',
 '_id',
 'version_methode_dpe',
 'geometry']

Il est trop coûteux de récupérer l'intégralité des données de DPE, et nous ne disposons de données météo que pour certaines communes. L'idée est donc de ne récupérer les données de DPE que pour les communes voulues et de les agréger dans une unique table de données.

In [72]:
liste_code_commune = filtre['INSEE_COM']
def gros_data_frame(liste_commune):
    DPE=pd.DataFrame(columns=[list(dpe)])
    for i in liste_commune: 
        code_commune=i
        size = 100
        api_root = "https://koumoul.com/data-fair/api/v1/datasets/dpe-france/lines"
        url_api = f"{api_root}?page=1&after=10&format=json&q_mode=simple&qs=code_insee_commune_actualise" + "%3A%22" + f"{code_commune}" + "%22" + f"&size={size}&select=" + "%2A&sampling=neighbors"
        DPE=pd.concat([DPE,get_dpe_from_url(url_api)])


big_data=gros_data_frame(liste_code_commune)
big_data

AttributeError: 'DataFrame' object has no attribute 'longitude'

In [200]:
import time
def api_ademe(codes_insee, nbr_page = 10):
    
    base_url = "https://koumoul.com/data-fair/api/v1/datasets/dpe-france/lines?page="
    
    for code_insee in codes_insee :
        for i in range(1,10*nbr_page + 2,10):
            time.sleep(5)
            page_init = i
            page_suiv = i + 9
    
            url = base_url + f"{page_init}" + "&after=" + f"{page_suiv}" + "&format=json&qs=code_insee_commune_actualise:" + code_insee
            req_api = requests.get(url)
            df = pd.json_normalize(req_api.json()['results'])
            data.append(df[~df['geo_adresse'].isna()])

    return pd.concat(data,axis=0,ignore_index=True)

filtre_conso_2021['Villes']

In [64]:
def get_json_results(data, code_commune):
    updated_data = data.copy()
    api_root = "https://koumoul.com/data-fair/api/v1/datasets/dpe-france/lines"
    url_api = f"{api_root}?page=1&after=10&format=json&q_mode=simple&qs=code_insee_commune_actualise" + "%3A%22" + f"{code_commune}" + "%22" + "&select=" + "%2A&sampling=neighbors"
    req=requests.get(url_api)
    wb=req.json()
    updated_data += wb['results']
    return updated_data

def api_ademe(codes_insee):
    data=[]
    for ville in codes_insee :
        data=get_json_results(data, ville)
    #return data
    df = pd.json_normalize(data)
    #return data
    dpe = gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(df.longitude, df.latitude), crs = 4326)
    dpe = dpe.dropna(subset = ['longitude', 'latitude'])
    return dpe



In [65]:
api=api_ademe(liste_communes)
api

classe_consommation_energie tr001_modele_dpe_type_libelle  \
0                              F                      Location   
1                              D                         Vente   
2                              D                      Location   
3                              N                         Vente   
4                              D                         Vente   
...                          ...                           ...   
3616                           F                         Vente   
3617                           B                          Neuf   
3619                           D                         Vente   
3620                           B                          Neuf   
3621                           N                         Vente   

      annee_construction           _geopoint   latitude  \
0                   1974  50.105335,1.832982  50.105335   
1                   1948  50.095824,1.841555  50.095824   
2                   1982  50.109299,1.842122  50.109299   
3                   1974  50.109172,1.806312  50.109172   
4                   1974  50.108204,1.837013  50.108204   
...                  ...                 ...        ...   
3616                1981   45.70891,4.585596  45.708910   
3617                2016  45.709107,4.586496  45.709107   
3619                1948  45.718229,4.605208  45.718229   
3620                2019  45.705667,4.591846  45.705667   
3621                1940  45.708522,4.590945  45.708522   

      surface_thermique_lot      _i tr002_type_batiment_description  \
0                    100.00      54             Maison Individuelle   
1                     90.00      72             Maison Individuelle   
2                     42.80      78                        Logement   
3                    182.05     220             Maison Individuelle   
4                    146.80     230                        Logement   
...                     ...     ...                             ...   
3616                 135.00  111488             Maison Individuelle   
3617                 122.22  118981             Maison Individuelle   
3619                 115.00  219700             Maison Individuelle   
3620                 133.43  259989             Maison Individuelle   
3621                  47.00  286355             Maison Individuelle   

                                   geo_adresse   _rand  ...  \
0         12 Place Max Lejeune 80100 Abbeville  786615  ...   
1           20 Rue de la Plume 80100 Abbeville   27716  ...   
2     2 Avenue Aristide Briand 80100 Abbeville  221098  ...   
3      246 Chaussée de Rouvroy 80100 Abbeville  680450  ...   
4          27 Chaussée du Bois 80100 Abbeville  531806  ...   
...                                        ...     ...  ...   
3616        7 Montée des Bruyères 69510 Yzeron   67993  ...   
3617          Montée des Bruyères 69510 Yzeron  326477  ...   
3619      2294 Route de la Brally 69510 Yzeron  148079  ...   
3620             Chemin du Planil 69510 Yzeron  808398  ...   
3621         4 Montée de l'Eglise 69510 Yzeron  230294  ...   

     classe_estimation_ges  version_methode_dpe  nom_methode_dpe  \
0                        D                V2012        3CL - DPE   
1                        E                V2012    FACTURE - DPE   
2                        B                V2012        3CL - DPE   
3                        N                V2012        3CL - DPE   
4                        E                V2012        3CL - DPE   
...                    ...                  ...              ...   
3616                     G                V2012        3CL - DPE   
3617                     A                V2012    FACTURE - DPE   
3619                     A                V2012    FACTURE - DPE   
3620                     A                V2012    FACTURE - DPE   
3621                     N                V2012        3CL - DPE   

     tv016_departement_code consommation_energie date_etablissement_dpe  \
0

In [66]:
api["code_insee_commune_actualise"].nunique()

316

### 1.6 Fusion des tableaux

Il ne nous reste enfin plus qu'à fusionner les tableaux pour obtenir dans une même ligne toutes les données qui nous semblent importantes, à savoir les données de consommations, les données DPE et les données météorologiques annuelles.

In [ ]:
for mot in big_data['geo_adresse'] :
    mot.upper()

data_dpe_2018=filtre_conso_2018.merge(big_data, left_on='Adresse normalisée', right_on='geo_adresse')
data_dpe_2018

In [ ]:
data_reg_2018=data_dpe_2018.merge(meteo_2018, left_on='NOM_M', right_on='Villes')